src/GPS.cpp

```diff
@@ -0,0 +1,114 @@
+#include "GPS.hpp"
+#include <nmea0183/GPS.hpp>
+
+using namespace nmea0183;
+using namespace marnav;
+using namespace std;
+using namespace gps_base;
+
+GPS_SOLUTION_TYPES GPS::getPositionType(utils::optional<nmea::mode_indicator> mode)
+{
+    switch (*mode) {
+        case nmea::mode_indicator::invalid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::autonomous:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        case nmea::mode_indicator::differential:
+            return GPS_SOLUTION_TYPES::DIFFERENTIAL;
+        case nmea::mode_indicator::estimated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::manual_input:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::simulated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::data_not_valid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::precise:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        default:
+            return GPS_SOLUTION_TYPES::INVALID;
+    }
+}
+
+base::Time GPS::buildRockTime(utils::optional<nmea::time> const& optional_time,
+    utils::optional<nmea::date> const& optional_date)
+{
+    if (optional_date.has_value() && optional_time.has_value()) {
+        auto date = optional_date.value();
+        auto time = optional_time.value();
+        return base::Time::fromTimeValues(date.year() + 2000,
+            static_cast<int>(date.mon()),
+            date.day(),
+            time.hour(),
+            time.minutes(),
+            time.seconds(),
+            time.milliseconds(),
+            0);
+    }
+    else {
+        return base::Time::now();
+    }
+}
+
+Solution GPS::getSolution(nmea::rmc const& rmc, nmea::gsa const& gsa)
+{
+    Solution solution;
+    auto mode_indicator = rmc.get_mode_ind().value();
```

Comment:
shouldn't you check if it has value before calling .value()? or pass it directly to getPositionType and check there

Reply:
If it returns a null pointer at getPositionType it will return a invalid solution

Reply:
If it return a null pointer, at getPositionType will be returned a invalid solution

Reply:
does the switch work with null pointer? isn't it doing *(null pointer) going to make it explode?

Reply:
just add a test

Reply:
it is not really a pointer
![image](https://github.com/user-attachments/assets/d65bff45-51a5-4215-b461-cf923ef4049d)

Reply:
it is not really a pointer
![Uploading image.png…]()

Reply:
The documentation says acessing an optional without a value is undefined behavior, so its naturally a bad practice to leverage an undefined behavior in the code. It being undefined can result in a heisenbug when the mode is initialized as one of the not INVALID states.

Reply:
I agree with Kappes, you should add a has_value check in the if before the invalid check, it is zero effort and would make the code way safer

Reply:
just changed